TokenBucket.php

```diff
@@ -0,0 +1,135 @@
+    */
+   public function __construct($key, Backend $backend, TokenRate $rate) {
+      if (!is_string($key)) {
+         throw new InvalidArgumentException("identifier must be a string");
+      }
+
+      $this->key = $key;
+      $this->backend = $backend;
+      $this->rate = $rate;
+   }
+
+   /**
+    * Tries to consume a token, if a token can't be consumed, then a DateTime
+    * is returned for when the next token will be available.
+    *
+    * @return array index 0 being a boolean whether or not a token was consumed
+    *               index 1 being a timestamp of when there will be
+    *               enough tokens to consume the amount requested.
+    */
+   public function consume($amount) {
+      if (!is_int($amount)) {
+         throw new InvalidArgumentException("amount must be an int");
+      }
+
+      list($tokens, $lastConsume) = $this->getTokenCountHelper();
+      $tokens -= $amount;
+      if ($tokens < 0) {
+         return [false, $this->readyTime($amount, $tokens, $lastConsume)];
+      }
+
+      $now = new DateTime();
+      $this->storeTokens($tokens, $now->getTimestamp());
+      return [true, $now->getTimestamp()];
+   }
+
+   public function getTokenCount() {
+      list($tokens, $_) = $this->getTokenCountHelper();
+      return $tokens;
+   }
+
+   /**
+    * @return array index 0 the number of tokens in the bucket
+    *               index 1 the last consumtion of a token, this will be the
+    *               current time if there wasn't a stored bucket.
+    */
+   private function getTokenCountHelper() {
+      $now = (new DateTime());
+      $storedBucket = $this->backend->get($this->key);
+      if ($storedBucket === Backend::MISS) {
+         return [$this->rate->tokens, $now];
+      }
+
+      return [$this->updateTokens($storedBucket->getTokens(),
+       $storedBucket->getLastConsume()), $storedBucket->getLastConsume()];
+      }
+
+   /**
+    * Updates tokens to the amount that it should be after restoring the tokens
+    * that have been regenerated from the last consume to now.
+    */
+   private function updateTokens($tokens, DateTime $lastConsume) {
+      if (!is_int($tokens)) {
+         throw new InvalidArgumentException("tokens must be an int");
+      }
+
+      $timeLapse = $lastConsume->diff(new DateTime())->s;
+      if ($timeLapse == 0) {
+         return $tokens;
+      }
+
+      $tokens += floor($timeLapse * $this->rate->getRate());
+      // Don't go over int max if we're going to use this as an int.
+      $tokens = (int)min(PHP_INT_MAX, $tokens);
+      // don't go over maximum tokens.
+      return min($this->rate->tokens, $tokens);
+   }
+
+   /**
+    * Returns the DateTime when the amount requested will be available.
+    * if the amount requested is higher than the max of the token bucket then
+    * it will return null, since there will never be a ready time.
+    */
+   private function readyTime($amount, $tokens, $lastConsume) {
```

Comment:
for the first parameter, it's not always the max tokens, it is also the amount being consumed in another call site.

The other two are time sensative, so I decided to not store them as members because they'll probably be out of date, and are not worth keeps as state. I should probably write that down in a comment of the class.